Move var declaration near reference

afilina opened the pull request:
Fixes #3048

afilina commented on rules/solid/src/Rector/Variable/MoveVariableDeclarationNearReferenceRector.php:
```diff
@@ -0,0 +1,115 @@
+        $assignStatement = $assign->getAttribute(AttributeKey::CURRENT_STATEMENT);
+
+        $this->addNodeBeforeNode($assignStatement, $usageStatement);
+        $this->removeNode($assignStatement);
```
@TomasVotruba This node manipulation seems to have no effect in case `$parentScope` is a `Closure`. Everything else looks good in the debugger. Perhaps a bug in Rector?

TomasVotruba replied:
Some nodes cannot be removed just like that.

E.g. if you remove `$if->cond` from `if($someCond)`, the `If_` would collapse.

In non-simple cases like these, one needs to work/remove the propper node, e.g. `$if->cond = null` or `$this->remove($ifNode)`.

afilina replied:
Right, but that doesn't change the fact that this piece of code works fine on every scope except a closure. The code being manipulated is the same in every other way, so something about closures seem to cause unexpected problem.

TomasVotruba replied:
Agreed. The Closure needs to be handled in custom way here

afilina replied:
I really don't see how to do that if node changes to a closure aren't registered properly in Rector. Is there some sort of weird exclusion in some collectors or visitors somewhere? I don't have enough Rector knowledge to be able to debug something like this. `$this->removeNode` does nothing.

TomasVotruba replied:
It's hard to say without real code. No need to work with `removeNode()`, as it's designed only for simple cases with 1 way to solve. In your case there might be many ways you'd like to solve it.

In general, I'd look for a `Closure_` and its location file, e.g.

```php
$parent = $assign->getAttribute(AttributeKey::PARENT);
if ($parent instanceof Closure_) {
   $closureParent = $parent->getAttribute(AttributeKey::PARENT);

   // $closureParent - what type is it? is it argument? assign? 
   // there are various way to handle this; remove `$closureParent` or null one of $closureParent properties, etc.
}
```

Real minimal PHP code with the closure and the node you're trying to remove would help, e.g.

```php
 $value = function () { 
-   $item = 5;
 };
```

afilina replied:
The test case is in the PR already: rules/solid/tests/Rector/Variable/MoveVariableDeclarationNearReferenceRector/Fixture/variable_in_closure.php.inc

TomasVotruba replied:
Thanks!

Link to the file would save me few clicks. Moreover on phone, where this takes copy-pasting pro skills

TomasVotruba replied:
![image](https://user-images.githubusercontent.com/924196/96038739-e2894780-0e67-11eb-866a-a153844b6934.png)


So if you the removing works correctly. The adding node is broken :)

afilina commented on rules/solid/tests/Rector/Variable/MoveVariableDeclarationNearReferenceRector/Fixture/variable_with_usage_in_multiple_levels.php.inc:
```diff
@@ -0,0 +1,25 @@
+    if (mktime() === false) {
+        return $var;
+    }
+    echo $var;
```
@TomasVotruba Is there some kind of helper in Rector to assist in finding the correct location for the assign statement in this situation? It looks like that would be a giant thing to write from scratch.

TomasVotruba replied:
I'm not aware of it. This Rule is the only place such service would be used in

afilina replied:
Definitely not an "easy pick" then :)
I'd have to get back to it later then, because this opens up a lot of potential use cases that we'd need to write for this Rector.

TomasVotruba replied:
Depends on the scope you go for :) No need to go so deep on 1st iteration. 

I'd be happy to merge it as it is without thins one case and then have a new PR.

afilina replied:
I'm not comfortable merging as is because that would cause the Rector to make incorrect changes and break the logic.

TomasVotruba replied:
Agreed. Broken code is not our goal.

That's why we go for opt-in approach, and skip such case completely. Then cover them in next PR.

afilina replied:
But we need to detect these cases to know that we need to skip them. If you unleash it as-is, in this particular case, rector would blindly move the assignment to the wrong location, causing the code to explode in one of the execution paths.

TomasVotruba replied:
That's true, but that's on purpose. It's the fastest way to discover such bugs, before it goes to general public. We release early and often, so bleeding edge testing can discover these hots spots ASAP.

Also, Rector is not a package in vendor, where breaking has different role.

TomasVotruba commented on docs/rector_rules_overview.md:
````diff
@@ -14158,6 +14158,26 @@ Classes that have no children nor are used, should have abstract
+ }
+```
+
+<br><br>
````
This looks like manual... is this generated with `composer docs` command?

afilina replied:
I copied it from the section above.

TomasVotruba replied:
No need to. The `composer docs` generates this feil

TomasVotruba commented on rules/solid/tests/Rector/Variable/MoveVariableDeclarationNearReferenceRector/Fixture/property_in_method.php.inc:
```diff
@@ -0,0 +1,35 @@
+        }
+    }
+}
+
+?>
+-----
+<?php
+
+namespace Rector\SOLID\Tests\Rector\Variable\MoveVariableDeclarationNearReferenceRector\Fixture;
+
+class PropertyInMethod
+{
+    private $var;
+    function myMethod()
+    {
+        $this->var = 1;
+        if (mktime() === false) {
+            return $this->var;
+        }
+    }
+}
+
+?>
```
In case of no-change, this whole part can be removed.

TomasVotruba commented on rules/solid/tests/Rector/Variable/MoveVariableDeclarationNearReferenceRector/Fixture/variable_in_closure.php.inc:
```diff
@@ -0,0 +1,27 @@
+function variableInClosure()
+{
+    $closure = function () {
+        $var = 1;
```
So the problem is that removing assing does not remove this line?

TomasVotruba replied:
In case of stmts (each function like has many stmts), you might need to iterate these stmts and unset them.

I'll try it and commit it to the PR

afilina replied:
Both the removing of assign and the adding of the assign somewhere else doesn't do anything.

TomasVotruba commented:
I've fixed cs, rebased on `master` and somehow made the test pass

TomasVotruba commented:
Got it :) uff, nasty bug

TomasVotruba commented:
I've added you higher access to repository, so we can work on the same PR without rebase/upsync and other mess: https://github.com/rectorphp/rector/invitations

afilina commented:
> I've fixed cs, rebased on `master` and somehow made the test pass.
> 
> Github pushed a new branch. Could you enable external contributions to your PRs? I don't know if that will help though
> 
> Here is the pull-request: #4410

I was pretty sure that I enabled editing of my PR when I opened it. Not sure how GitHub handles this.
